### PR TITLE
Fix editable CMake prefix path

### PR DESCRIPTION
**Pull request by khlaifiabilel**
I reproduced this issue while testing a local editable PyTorch build: OpenReg could not locate the installed CMake configuration files. This fix makes the development setup more reliable for contributors building C++ extensions.

> **AI-assisted summary**
>
> Editable installs load Python modules from the source tree while compiled modules and package data remain in the physical installation directory. Deriving package paths from Python module `__file__` values therefore makes `torch.utils.cmake_prefix_path` point to a source directory without the installed CMake metadata.
>
> The OpenReg extension setup used the same assumption, preventing its nested CMake build from finding `TorchConfig.cmake`. Derive both paths from `torch._C.__file__`, which remains in the physical package directory for regular and editable installs.
>
> Add regression coverage for the public CMake prefix path.
>
> **Test plan**
>
> ```bash
> PATH="$PWD/venv/bin:$PATH" python test/test_utils.py TestCppExtensionUtils
> PATH="$PWD/venv/bin:$PATH" python test/run_test.py --openreg
> PATH="$PWD/venv/bin:$PATH" lintrunner -a
> ```
>
> This summary was prepared with an AI coding assistant. I reviewed the implementation, test results, and PR text and take responsibility for the contribution.

cc @malfet @zklaus

### CODE REVIEW

**Comment by khlaifiabilel** [comment]
I have verified the files impacted by this pull request. The modifications indeed concern the CMake build system and OpenReg integration; therefore, both module labels are accurately applied.

> AI-assisted routing suggestion: this PR changes general CMake package discovery and the OpenReg extension setup, so the relevant routing labels are `module: build` and `module: openreg`.

```
@pytorchbot label "module: build" "module: openreg"
```

**Review by Skylion007 — DISMISSED** [review]

File: test/cpp_extensions/open_registration_extension/torch_openreg/setup.py [hunk 1]
```diff
@@ -36,7 +36,7 @@ def get_pytorch_dir():
     os.environ["TORCH_DEVICE_BACKEND_AUTOLOAD"] = "0"
     import torch
 
-    return os.path.dirname(os.path.realpath(torch.__file__))
+    return os.path.dirname(os.path.realpath(torch._C.__file__))
```
Comment (malfet) [inline test/cpp_extensions/open_registration_extension/torch_openreg/setup.py:39]: Why make this change? It returns the same value, isn't it? (I have an editable setup)
```
% python -c "import torch; import os; print(os.path.dirname(os.path.realpath(torch.__file__)), os.path.dirname(os.path.realpath(torch._C.__file__)))"
/Users/malfet/git/pytorch/pytorch/torch /Users/malfet/git/pytorch/pytorch/torch
```

**Review by malfet — DISMISSED** [review]
Sure, not sure we want to introduce a new test, but why not

**Review by malfet — CHANGES_REQUESTED** [review]
@khlaifiabilel is there an issue that explains what this editable setup is? Because for me both `torch.__file__` and `torch._C.__file__`  are always located in the same folder
```
% python -c 'import torch;import os;print(torch.utils.cmake_prefix_path, os.path.join(os.path.dirname(torch._C.__file__), "share", "cmake"))'
/Users/malfet/git/pytorch/pytorch/torch/share/cmake /Users/malfet/git/pytorch/pytorch/torch/share/cmake
```

**Comment by janeyx99** [comment]
Thanks for your interest in submitting PRs to help our project. Same as before, to avoid you spending time preparing a PR for a feature we're not sure we want to add, our usual process is to discuss on an issue before doing the PR.

See https://github.com/pytorch/pytorch/wiki/The-Ultimate-Guide-to-PyTorch-Contributions for more details.
